gslib/commands/iam.py

```diff
@@ -363,8 +373,106 @@ def get_gcloud_storage_args(self):
       # Gsutil takes policy file then URLs, and gcloud storage does the reverse.
       self.args = url_strings + self.args[:1]
 
+    elif self.sub_command == 'ch':
+      return []  # Translation occurs in self.run_gcloud_storage.
+
     return super().get_gcloud_storage_args(gcloud_storage_map)
 
+  def _RaiseIfInvalidUrl(self, url):
+    if url.IsFileUrl():
+      error_msg = 'Invalid Cloud URL "%s".' % url.object_name
+      if set(url.object_name).issubset(set('-Rrf')):
+        error_msg += (
+            ' This resource handle looks like a flag, which must appear'
+            ' before all bindings. See "gsutil help iam ch" for more details.')
+      raise CommandException(error_msg)
+
+  def _run_gcloud_subprocess(self, args, stdin=None):
+    subprocess_envs = os.environ.copy()
+    subprocess_envs.update(self._translated_env_variables)
+
+    _, command = self._get_full_gcloud_storage_execution_information(args)
+    process = subprocess.run(
+        command,
+        env=subprocess_envs,
+        input=stdin,
+        stderr=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        text=True,
+    )
+    if process.returncode != 0 and not self.continue_on_error:
+      raise CommandException(process.stderr)
+    return process
+
+  def run_gcloud_storage(self):
+    if self.sub_command != 'ch':
+      return super().run_gcloud_storage()
+
+    self.ParseSubOpts()
+    bindings_tuples, patterns = self._GetSettingsAndDiffs()
+    resource_type = self._get_resource_type(patterns)
+
+    list_settings = []
+    if self.recursion_requested:
+      list_settings.append('-r')
+
+    return_code = 0
+    for url_pattern in patterns:
+      self._RaiseIfInvalidUrl(url_pattern)
+
+      if resource_type == 'objects':
+        ls_process = self._run_gcloud_subprocess(['alpha', 'storage', 'ls'] +
+                                                 list_settings +
+                                                 [str(url_pattern)])
+        if ls_process.returncode:
+          return_code = 1
+          continue
+
+        urls = ls_process.stdout.strip('\n').split('\n')
+      else:
+        urls = [str(url_pattern)]
+
+      # With the -r flag, container headings are printed ending in '/:'.
+      # All headings other than the first are preceded by a new line.
+      # These need to be skipped, without skipping objects that end in
+      # '/:'.
+      next_url_could_be_container_line = True
```

Comment:
This would be more robust if it was based on reading object attributes from `ls --json` or `objects list` instead of determining data types from output formatting.

Reply:
Agreed! `objects list` doesn't quite work since it only lists the contents of directories with a wildcard, so it functions differently than recursive listing:

```
$ gcloud storage objects list gs://bucket/dir
$ gcloud storage objects list gs://bucket/dir/*
...
name: dir/file
...
```

ls --json seems to work though!